### ISSUE TITLE
Immutable DynamoDb Value.Style Suggestions + Documentation Update

### ISSUE THREAD

**Issue by swar8080** [immutables/immutables#1478]
Hello, i'm looking for feedback on how to use `@Value.Style` when working with the V2 AWS SDK's [DynamoDB Enhanced Client](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/dynamodb-enhanced-client.html). 

The current immutables [documentation](https://immutables.github.io/dynamodb.html) is for `DynamoDbMapper` from v1 of the aws sdk, which was replaced with the enhanced client in v2. I can update the documentation if you think the v2 `@Value.Style` pattern makes sense after getting your feedback?

### Possible Style for immutable DynamoDb Enhanced Client items 
The enhanced client can [support immutable items that use builders](https://github.com/aws/aws-sdk-java-v2/tree/master/services-custom/dynamodb-enhanced#working-with-immutable-data-classes) that meet this criteria:
> 1. Every method on the immutable class that is not an override of Object.class or annotated with @DynamoDbIgnore must be a getter for an attribute of the database record.
> 2. Every getter in the immutable class must have a corresponding setter on the builder class that has a case-sensitive matching name.
> 3. EITHER: the builder class must have a public default constructor; OR: there must be a public static method named 'builder' on the immutable class that takes no parameters and returns an instance of the builder class.
> 4. The builder class must have a public method named 'build' that takes no parameters and returns an instance of the immutable class.

I was able to meet the criteria using this style:
```
@Value.Style(
    // Needed to remove the from(*) builder methods since the SDK thinks "from" is a column in the table.
    // Is there a way to remove the "from" methods without the other consequences of strictBuilder?  
    strictBuilder = true,
    
    //Same as above, the SDK thinks the addAll* methods correspond to columns in the table
    builtinContainerAttributes = false,
    
    //Same as above, the SDK thinks the copy methods correspond to columns in the table
    defaults = @Value.Immutable(copy = false),

    // All the valid SDK annotations that can be added to item classes and their getters
    passAnnotations = {
      DynamoDbImmutable.class,
      DynamoDbAttribute.class,
      DynamoDbConvertedBy.class,
      DynamoDbFlatten.class,
      DynamoDbIgnore.class,
      DynamoDbIgnoreNulls.class,
      DynamoDbPartitionKey.class,
      DynamoDbPreserveEmptyObject.class,
      DynamoDbSecondaryPartitionKey.class,
      DynamoDbSecondarySortKey.class,
      DynamoDbSortKey.class,
      DynamoDbUpdateBehavior.class,
      DynamoDbAtomicCounter.class,
      DynamoDbAutoGeneratedTimestampAttribute.class,
      DynamoDbVersionAttribute.class,
    })
public @interface DynamoDbStyle {}
```

And an example usage:
```
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbImmutable;
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbPartitionKey;
import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbSortKey;

@DynamoDbStyle
@Value.Immutable
@DynamoDbImmutable(builder = ImmutableCustomer.Builder.class)
public interface Customer {

  @DynamoDbPartitionKey
  String getPk();

  @DynamoDbSortKey
  String getSk();

  String getName();

  //Nullable is used here since the SDK doesn't have built-in support for Optionals.
  //Immutables doesn't support nullable collections so the list of phone numbers needs to be wrapped in a POJO.
  @Nullable
  PhoneNumbers getPhoneNumbers();

  class PhoneNumbers {
     private List<String> phoneNumbers;
     ...
  }
}
```

### Possible Style for updating client items 

Sometimes when updating an item you only want to modify specific field values. For example, when atomically incrementing a column you don't want to submit other field values since they could be out-of-date. 

However, the SDK takes the entire item as input with it's update API, but allows you to set field values to `null` to indicate they shouldn't be modified. So it'd be convenient to re-use the immutable item class's interface for updates without removing the checks on required fields.

So one approach is something like:

```
@Value.Style(
    // Disable required field checks
    validationMethod = ValidationMethod.NONE,

    // Same style as above
    strictBuilder = true,
    builtinContainerAttributes = false,
    defaults = @Value.Immutable(copy = false),
    passAnnotations = {
      DynamoDbImmutable.class,
      ...
    })
@Value.Immutable
@DynamoDbImmutable(builder = ImmutableCustomerUpdate.Builder.class)
public interface CustomerItemUpdate extends Customer {}
```

### Full Example

```
// Insert an a Customer item
TableSchema<ImmutableCustomer> customerSchema =
    TableSchema.fromImmutableClass(ImmutableCustomer.class);
DynamoDbTable<ImmutableCustomer> customerTable =
    enhancedClient.table("customers", customerSchema);

ImmutableCustomer customer =
    ImmutableCustomer.builder()
      .pk(UUID.randomUUID().toString())
      .sk(UUID.randomUUID().toString())
      .name("customer1")
      .phoneNumbers(new PhoneNumbers(Arrays.asList("123-456-7890", "111-111-1111")))

customerTable.putItem(customer);

// Update only the customer's phone numbers, not their name
TableSchema<ImmutableCustomerUpdate> updateSchema =
    TableSchema.fromImmutableClass(ImmutableCustomerUpdate.class);
DynamoDbTable<ImmutableCustomerUpdate> customerUpdate =
    enhancedClient.table("customers", updateSchema);

ImmutableCustomerUpdate update =
  ImmutableCustomerUpdate.builder()
      .pk(customer.getPk())
      .sk(customer.getSk())
      // The name field doesn't need to be provided (even though it's required during creation)
      .phoneNumbers(Arrays.asList("000-0000-0000"))
      .build()

customerUpdate.updateItem(builder -> builder.item(update).ignoreNulls(true).build());
```

**Comment by elucash** [comment]
```java
// this works when jdkOnly = true or when no guava found on classpath
// @AllowNulls - your own annotations with same simple name
@Nullable List<@AllowNulls String> attr();
// can be type_use or a method level annotation
```

will see if it's easy to suppress `from` method (I've though it was)

Updating the documentation would be very welcome (`src` branch in [`immutables.github.io`](https://github.com/immutables/immutables.github.io) repo)

Guess we need to find / fix some more workarounds. What's left except that `from` method? Handling partials (partially initialized) is hard in any setup. Have you tried to messing around with `Modifiable` (builder-in-disguise, search issues for more info) ?

**Comment by swar8080** [comment]
> will see if it's easy to suppress from method (I've though it was)

That'd be great, I didn't see anything when reading through the `@Value` javadoc. If strict builders are the only approach then i'd be interested in implementing an option to suppress `from`?


> // this works when jdkOnly = true or when no guava found on classpath
// @AllowNulls - your own annotations with same simple name
@Nullable List<@AllowNulls String> attr();

`@Nullable` collection attributes are actually working perfectly when testing, even with guava collections. I vaguely remember it causing problems a few weeks ago, but maybe that was with a different `@Value.Style` I was experimenting with. Do you remember the limitation that led to the documentation "[Nullable collections and other special types are not supported](http://immutables.github.io/immutable.html#nullable-attributes)"?

Although maybe `jdkOnly = true` makes sense either way for this style since dynamo supports null collection elements?


> Handling partials (partially initialized) is hard in any setup. Have you tried to messing around with Modifiable (builder-in-disguise, search issues for more info) ?

I got partials working by extending the non-partial interface and using the style option of `validationMethod = ValidationMethod.NONE`. 
Note that this generate a warning but it hasn't caused any issues:

```
[WARNING] /Users/sswartz/workspace/dynamodb-enhance-client-scratch/src/main/java/org/swar8080/dynamodbenhancedscratch/immutables/CustomerUpdate.java:[9,8] 

(immutables:subtype) Should not inherit org.swar8080.dynamodbenhancedscratch.immutables.Customer which is a value type itself. Avoid extending from another abstract value type. 

Better to share common abstract class or interface which are not carrying @Immutable annotation. If still extending from immutable abstract type be ready to face some incoherences in generated types.
```

In the new documentation we could do what's recommended by that warning, but it'd be more convenient having 2 classes instead of 3

**Comment by swar8080** [comment]
Let me know if it's worth adding the documentation with the current approach and strict builders limitation. I can update it if/when I add an option to skip generating the `from` method.

**Comment by elucash** [comment]
Yes, documentation about styles and usage are very welcome, definitely it needs an update. Feel free to PR `src` branch in [immutables.github.io](https://github.com/immutables/immutables.github.io) repo

I'm looking into `from` method. If this will work, in `2.10.0` we will be able to use `Style(from="",` to disable `from` method, by using empty string as a naming pattern. 

BTW `@SuppressWarnings("immutables:subtype")` can be put on a class or package to silence "Should not inherit.."

**Comment by elucash** [comment]
> Do you remember the limitation that led to the documentation "[Nullable collections and other special types are not supported](http://immutables.github.io/immutable.html#nullable-attributes)"?

May implementations of immutable collections, including Guava's and JDK 9 immutable collections do not support nulls as elements. Even with `jdkOnly=true` we ban nulls (unless `@AllowNulls` or `@SkipNulls`). Overall, the problem with `null` is that it is not typed (`@Nullable` annotations and `Optional` can be used as an improvement, but still) and that developers, as result, overuse `null` in the wrong contexts, most of the time where it's not intended. I don't think that the fact that some concurrent data-structures use null for special things (and therefore do not allow null as values) is something which should be overly important, I think it's just something implementation specific and not hugely relevant in the art of avoiding `null`s. The software quality improves significantly when majority/all of developers no longer use null for arbitrary purposes (but don't ask me to provide research paper or a prove this is as a theorem, for me this is just certain observation).
I even think that it's not an overkill to use smth like `List<Optional<T>>` where this is what actually was meant - a list of elements where specific positions are absent. For the `Map` and `Set` even that is barely makes sense, there is no "positions", so something is more naturally is either included in set or not, either have value (for a key) or not, the need for `null`(or `Optional.empty()`) as a set element or as a value of a key is very rarely justifiable. So when I see in data/JSON something like this `[null, "a", null, null, "bc"]`  in 99% of cases this is bad data design or just programming error (silent errors resulting in undefined result which somehow turns into `null`), I do understand that something this is necessary, that's why we can use `List<Optional<T>>` or `List<@AllowNulls T>` and hopefully it will be documented why it's necessary

**Comment by swar8080** [comment]
> I'm looking into from method. If this will work, in 2.10.0 we will be able to use Style(from="", to disable from method, by using empty string as a naming pattern.


Alright, i'll open a MR for the updated dynamodb documentation that assumes `Style(from=""` will work once 2.10 is released

---

For the Nullable collection documentation, I was interpreting it as saying the collection itself can't be nullable, like `@Nullable List<String> foo`, but so far that hasn't caused me any problems. In the dynamo documentation I won't mention that as a limitation then

And also it sounds like the dynamodb documentation shouldn't mention `jdkOnly=true`, and instead leave it up to people to add if they want. I also think null collection elements are a code smell

**Comment by elucash** [comment]
yes, these are separate concerns, whether collection itself can be absent `@Nullable`, or what's happening with its elements (`@AllowNulls`/`@SkipNulls`). Yep, the said element annotations and `jdkOnly` can be irrelevant. 
Having collection itself as a `@Nullable` will prevent defaulting it to an empty collection, which is sometimes is ok and sometimes is not, so having collection nullable is a useful tool.
